core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -3774,11 +3780,136 @@ protected void validateJoin(SqlJoin join, SqlValidatorScope scope) {
+  /**
+   * Shuttle which determines whether all SqlCalls that are
+   * comparisons are comparing columns from both namespaces.
+   * The shuttle will throw an exception if that happens.
+   * If it returns all SqlCalls have the expected shape.
+   */
+  private class CompareFromBothSides extends SqlShuttle {
+    final SqlValidatorScope scope;
+    final SqlValidatorCatalogReader catalogReader;
+    final Resources.ExInst<SqlValidatorException> exception;
+
+    private CompareFromBothSides(
+        SqlValidatorScope scope,
+        SqlValidatorCatalogReader catalogReader,
+        Resources.ExInst<SqlValidatorException> exception) {
+      this.scope = scope;
+      this.catalogReader = catalogReader;
+      this.exception = exception;
+    }
+
+    @Override public @Nullable SqlNode visit(final SqlCall call) {
+      SqlKind kind = call.getKind();
+      if (SqlKind.COMPARISON.contains(kind)) {
+        assert call.getOperandList().size() == 2;
+
+        boolean leftFound = false;
+        boolean rightFound = false;
+        // The two sides of the comparison must be from different tables
+        for (SqlNode operand : call.getOperandList()) {
+          if (!(operand instanceof SqlIdentifier)) {
+            throw newValidationError(call, this.exception);
+          }
+          // We know that all identifiers have been expanded by the caller,
+          // so they have the shape namespace.field
+          SqlIdentifier id = (SqlIdentifier) operand;
+          final SqlNameMatcher nameMatcher = catalogReader.nameMatcher();
+          final SqlValidatorScope.ResolvedImpl resolved = new SqlValidatorScope.ResolvedImpl();
+          // Lookup just the first component of the name
+          scope.resolve(id.names.subList(0, id.names.size() - 1), nameMatcher, false, resolved);
+          SqlValidatorScope.Resolve resolve = resolved.only();
+          int index = resolve.path.steps().get(0).i;
+          if (index == 0) {
+            leftFound = true;
+          }
+          if (index == 1) {
+            rightFound = true;
+          }
+
+          if (!leftFound && !rightFound) {
+            throw newValidationError(call, this.exception);
+          }
+        }
+        if (!leftFound || !rightFound) {
+          // The comparison does not look at both tables
+          throw newValidationError(call, this.exception);
+        }
+      }
+      return super.visit(call);
+    }
+  }
+
+  /**
+   * Shuttle which determines whether an expression is a simple conjunction
+   * of equalities. */
+  private static class ConjunctionOfEqualities extends SqlShuttle {
+    boolean illegal = false;
+
+    @Override public @Nullable SqlNode visit(final org.apache.calcite.sql.SqlCall call) {
+      SqlKind kind = call.getKind();
+      if (kind != SqlKind.AND && kind != SqlKind.EQUALS) {
+        illegal = true;
+      }
+      if (kind == SqlKind.AND) {
+        List<SqlNode> operands = call.getOperandList();
+        for (SqlNode operand : operands) {
```

Comment:
It seems that there might still be an issue here. If the condition is a.id = b.id AND a.id2 = b.id2 AND a.id3 = b.id3, this might not pass as expected because the conditions in SqlNode are not in the same operandList.

Reply:
In Calcite AND can have N operands and I believe that there is a pass prior to this which will make sure that a sequence of AND operators is flattened into a single operator.

Reply:
It look like that is only true for RexNodes, so you are right, I will have to handle this case too.

Reply:
I have fixed this case and added a test. Great observation.

Reply:
Looks good✅